### DISCUSSION TITLE
How to use custom python packages for android?

### DISCUSSION THREAD

**Discussion by wschume** [flet-dev/flet#2300] [Q&A] [answered]
### Question

In the [documentation](https://flet.dev/docs/guides/python/packaging-app-for-distribution#native-python-packages-for-android-and-ios) it is mentioned that "Native Python packages for Android and iOS" require to use `serious_python`.
I would like to know if

1. Which python packages are considered "Native". Packages like math, shutil, subprocess etc? Or also Packages without any C, C++, Fortran etc. libraries like PyYAML, requests, etc?
2. I would like to access a server from my Android application. Is it possible to access a FastAPI Server or something similar from Android without the use of `serious_python`? I would like to make use of my backend in this form.

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by FeodorFitsner** [comment]
1. "Native" packages are packages partially (or fully) implemented in either C, C++, Rust or other language yielding "native" code, etc., i.e. having `.pyd` (Windows) or `.so` (macOS, Linux) files inside their wheels. I don't know if "native" is the correct word for such packages (another way to say that is "non-pure" packages) - I'm open to other ideas.
2. Yes, you can access FastAPI server or any other service from your Flet app. It's just Python.

**Reply by wschume** [reply]
@FeodorFitsner Thank you for your reply. Regarding to

1. I thought that you were referring to pure Python packages when using the term native. For me, a native Python package is something like sys, os or re. Packages that are "built-in" a Python deployment. Pure Python packages are packages that do not rely on any sort of binaries and use only python files. But I understand, that this definition is purely subjective.

2. Is your statement still valid, when trying to access a FastAPI server from a built Flet application (using `flet build`). I tried web and Android (and I think also Linux), and I was unable to access the FastAPI server with using either requests (a pure Python package) as well as urllib3 (a native Python package). I would love to be able to send HTTP requests from built Flet apps. Could you point me to a document/example where this is demonstrated? Or is this feature on the road map? 

Thank you for providing such an awesome package and the time you put into it!

**Reply by goncaze** [reply]
It appears that you are referring to the internal functions as an example of pure python. Example: abs(), dict(), help(), divmod(), id(), object(), classified(), ascii(), enumerate(). Thus, numpy can be an example of non-pure python, since it is composed of Python
61.1%, C 35.6%, C++ 1.6%, Cython 0.9%, Meson 0.5%, Fortran 0.2%, Other 0.1%.

**Reply by wschume** [reply]
@goncaze I agree, but I don't only mean the built-in functions like the one you listed, but also the modules.

**Comment by FeodorFitsner** [comment]
1. I agree, we have to "standardize" terminology across various sources. OK, "pure" Python package is the package that is entirely implemented in Python, without parts written in C or other languages. "Native" or "built-in" Python package is a package that is a part of Python standard library, i.e. comes bundled with a Python installation. How would you call packages with parts written in C, Rust or other language: "non-pure" package? Do we have a common name for those?
2. Have you tried https://flet.dev/docs/guides/python/packaging-app-for-distribution#ssl-fix-for-ios-and-android ?

**Comment by wschume** [comment, marked as answer]
I gave accessing a FastAPI server from Flet another try and it now worked like a charm. With the following FastAPI server:

```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def read_root():
    return {"Hello": "World"}
```

I was able to access the server with the following script:

```python
import requests
import flet as ft


async def main(page: ft.Page):
    page.title = "Flet counter example"
    page.vertical_alignment = ft.MainAxisAlignment.CENTER

    my_text = ft.Text(value="START")

    async def on_click(e):
        r = requests.get("http://<ip>:<port>/")

        my_text.value = r.content.decode("utf-8")
        await page.update_async()

    await page.add_async(
        ft.Row(
            [
                ft.ElevatedButton(text="Click", on_click=on_click),
                my_text,
            ],
            alignment=ft.MainAxisAlignment.CENTER,
        )
    )


ft.app(target=main)
```

I am unsure, what went wrong before. But I am very happy I can now build true android apps using python.
